pkg/flowaggregator/clickhouseclient/clickhouseclient.go

```diff
@@ -0,0 +1,511 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clickhouseclient
+
+import (
+	"database/sql"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ClickHouse/clickhouse-go"
+	"github.com/gammazero/deque"
+	ipfixentities "github.com/vmware/go-ipfix/pkg/entities"
+	"k8s.io/klog/v2"
+)
+
+const (
+	maxQueueSize   = 1 << 19 // 524288. ~500MB assuming 1KiB per record
+	commitInterval = 1 * time.Second
+	insertQuery    = `INSERT INTO flows (
+                   flowStartSeconds,
+                   flowEndSeconds,
+                   flowEndSecondsFromSourceNode,
+                   flowEndSecondsFromDestinationNode,
+                   flowEndReason,
+                   sourceIP,
+                   destinationIP,
+                   sourceTransportPort,
+                   destinationTransportPort,
+                   protocolIdentifier,
+                   packetTotalCount,
+                   octetTotalCount,
+                   packetDeltaCount,
+                   octetDeltaCount,
+                   reversePacketTotalCount,
+                   reverseOctetTotalCount,
+                   reversePacketDeltaCount,
+                   reverseOctetDeltaCount,
+                   sourcePodName,
+                   sourcePodNamespace,
+                   sourceNodeName,
+                   destinationPodName,
+                   destinationPodNamespace,
+                   destinationNodeName,
+                   destinationClusterIP,
+                   destinationServicePort,
+                   destinationServicePortName,
+                   ingressNetworkPolicyName,
+                   ingressNetworkPolicyNamespace,
+                   ingressNetworkPolicyRuleName,
+                   ingressNetworkPolicyRuleAction,
+                   ingressNetworkPolicyType,
+                   egressNetworkPolicyName,
+                   egressNetworkPolicyNamespace,
+                   egressNetworkPolicyRuleName,
+                   egressNetworkPolicyRuleAction,
+                   egressNetworkPolicyType,
+                   tcpState,
+                   flowType,
+                   sourcePodLabels,
+                   destinationPodLabels,
+                   throughput,
+                   reverseThroughput,
+                   throughputFromSourceNode,
+                   throughputFromDestinationNode,
+                   reverseThroughputFromSourceNode,
+                   reverseThroughputFromDestinationNode) 
+                   VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 
+                           ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`
+)
+
+type ClickHouseExportProcess struct {
+	db        *sql.DB
+	dsn       string
+	deque     *deque.Deque
+	mutex     sync.RWMutex
+	queueSize int
```

Comment:
After looking how deque, mutex, and queueSize are used, I'm wondering if this can achieved by a channel in a lock free way and avoid introducing the library:
1. Use a buffered channel with specific size to hold the queued items.
2. If the channel is full, drop the latest one by using `select` statement when sending the item to the channel (or is it a must to drop the oldest one? I thought it doesn't make much difference as the oldest and latest is relative).
3. In batchCommitAll, it collects all records that currently in the channel to a slice and try to commit it. If it fails, store the slice to a member variable `ch.lastUndeliveredItems` and retry it first in next round.

Reply:
Thanks Quan for your inputs. Addressed all the others. 
For this one, essentially two queues are formed. The following challenges exists here:

1. we no longer have a strict upper bound of the queue size of pending-commit records. we surely can limit the size of the buffered channel, but the length of `lastUndeliveredItems` will be dynamic with each commit cycle, and we can't change the size of channel based on length of slice.
2. when chan is full, and `lastUndeliveredItems` is non-empty, we should drop the ones in the slice first, as those are cached much earlier as we had at least tried once committing them. One straightforward approach would be to pop the head of chan to the slice. however, without a lock, there will for sure be racing conditions between the cacher and comitter. 

What do you think? Thanks.